// flatlist-react/lib/utils/getType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var typesMap = {
    array: 'ARRAY',
    boolean: 'BOOLEAN',
    function: 'FUNCTION',
    map: 'MAP',
    null: 'NULL',
    number: 'NUMBER',
    object: 'OBJECT',
    set: 'SET',
    string: 'STRING',
    symbol: 'SYMBOL',
    undefined: 'UNDEFINED',
    weak_map: 'WEAK_MAP',
    weak_set: 'WEAK_SET',
};
exports.types = Object.values(typesMap)
    .reduce(function (obj, type) {
    obj[type] = type;
    return obj;
}, {});
var getType = function (x) {
    var type = typeof x;
    switch (type) {
        case 'number':
        case 'string':
        case 'boolean':
        case 'undefined':
        case 'symbol':
        case 'function':
            return typesMap[type];
        default:
            return x === null ? typesMap.null :
                (x instanceof Set) ? typesMap.set :
                    (x instanceof WeakSet) ? typesMap.weak_set :
                        (x instanceof Map) ? typesMap.map :
                            (x instanceof WeakMap) ? typesMap.weak_map :
                                Array.isArray(x) ? typesMap.array :
                                    typesMap.object;
    }
};
exports.default = getType;
